src/engine/source/_builder/environment.hpp

```diff
@@ -0,0 +1,331 @@
+               const std::string& graphName,
+               Asset::Type type)
+    {
+        auto& graph = m_graphs[graphName];
+        for (auto& [name, json] : assetsDefinitons)
+        {
+            // Build Asset object and insert
+            auto asset = std::make_shared<Asset>(json, type);
+            m_assets.insert(std::make_pair(name, asset));
+            graph.addNode(name, asset);
+            if (asset->m_parents.empty())
+            {
+                graph.addEdge(graph.root(), name);
+            }
+            else
+            {
+                for (auto& parent : asset->m_parents)
+                {
+                    graph.addEdge(parent, name);
+                }
+            }
+        }
+    }
+
+    void addFilters(const std::string& graphName)
+    {
+        auto& graph = m_graphs[graphName];
+        for (auto& [name, asset] : m_assets)
+        {
+            if (asset->m_type == Asset::Type::FILTER)
+            {
+                for (auto& parent : asset->m_parents)
+                {
+                    if (graph.hasNode(parent))
+                    {
+                        graph.injectNode(name, asset, parent);
+                    }
+                }
+            }
+        }
+    }
+
+public:
+    Environment() = default;
+    template<typename T>
+    Environment(std::string name, const Json& jsonDefinition, T catalog)
+        : m_name {name}
+    {
+        auto envObj = jsonDefinition.getObject();
+
+        // Filters are not graphs, its treated as a special case.
+        // We just add them to the asset map and then inject them into each
+        // graph.
+        auto filtersPos = std::find_if(
+            envObj.begin(),
+            envObj.end(),
+            [](auto& tuple) { return std::get<0>(tuple) == FILTERS; });
+        if (filtersPos != envObj.end())
+        {
+            auto filtersList = std::get<1>(*filtersPos).getArray();
+            std::transform(
+                filtersList.begin(),
+                filtersList.end(),
+                std::inserter(m_assets, m_assets.begin()),
+                [&](auto& json)
+                {
+                    return std::make_pair(
+                        json.getString(),
+                        std::make_shared<Asset>(
+                            catalog.getAsset(1, std::string(json.getString())),
+                            Asset::Type::FILTER));
+                });
+            envObj.erase(filtersPos);
+        }
+
+        // Build graphs
+        for (auto& [name, json] : envObj)
+        {
+            auto assetNames = json.getArray();
+
+            m_graphs.insert(
+                std::make_pair<std::string,
+                               Graph<std::string, std::shared_ptr<Asset>>>(
+                    std::string {name},
+                    {std::string(name + "Input"),
+                     std::make_shared<Asset>(name + "Input",
+                                             getAssetType(name))}));
```

Comment:
could we get the asset type first and store it in a variable inside the for loop in order to avoid calling it 3 times in the same block ?